src/crimson/os/seastore/onode_manager/staged-fltree/node_extent_visitor.h

```diff
@@ -0,0 +1,233 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include "node_extent_manager.h"
+#include "node_delta_recorder.h"
+#include "node_layout_replayable.h"
+
+namespace crimson::os::seastore::onode {
+
+template <typename FieldType, node_type_t NODE_TYPE>
+class DeltaRecorderT final: public DeltaRecorder {
+ public:
+  using layout_t = NodeLayoutReplayableT<FieldType, NODE_TYPE>;
+  using position_t = typename layout_t::position_t;
+  using StagedIterator = typename layout_t::StagedIterator;
+  using value_t = typename layout_t::value_t;
+  static constexpr auto FIELD_TYPE = layout_t::FIELD_TYPE;
+
+  ~DeltaRecorderT() override = default;
+
+  template <KeyT KT>
+  void encode_insert(
+      const full_key_t<KT>& key,
+      const value_t& value,
+      const position_t& insert_pos,
+      const match_stage_t& insert_stage,
+      const node_offset_t& insert_size) {
+    // TODO encode to encoded
+  }
+
+  void encode_split(
+      const StagedIterator& split_at,
+      const char* p_start) {
+    // TODO encode to encoded
+  }
+
+  template <KeyT KT>
+  void encode_split_insert(
+      const StagedIterator& split_at,
+      const full_key_t<KT>& key,
+      const value_t& value,
+      const position_t& insert_pos,
+      const match_stage_t& insert_stage,
+      const node_offset_t& insert_size,
+      const char* p_start) {
+    // TODO encode to encoded
+  }
+
+  void encode_update_child_addr(
+      const laddr_t new_addr,
+      const laddr_packed_t* p_addr,
+      const char* p_start) {
+    // TODO encode to encoded
+  }
+
+  static DeltaRecorderURef create() {
+    return std::unique_ptr<DeltaRecorder>(new DeltaRecorderT());
+  }
+
+ private:
+  DeltaRecorderT() = default;
+  node_type_t node_type() const override { return NODE_TYPE; }
+  field_type_t field_type() const override { return FIELD_TYPE; }
+  void apply_delta(ceph::bufferlist::const_iterator& delta,
+                   NodeExtentMutable& node) override {
+    assert(is_empty());
+    // TODO decode and apply
```

Comment:
I guess replay support is still TODO?

Reply:
The architecture is ready. It will load the typed `NodeLayoutReplayableT` (https://github.com/cyx1231st/ceph/blob/wip-seastar-onode-tree-upstream/src/crimson/os/seastore/onode_manager/staged-fltree/node_layout_replayable.h#L15) and call its methods to replay the extent after the parameters are decoded from the delta. The only thing left is to encode,  decode parameters and unit tests. I'll get back to them after fixing more urgent messenger issues.